Add SOP for schema changes (DataBiosphere/azul#2884) (#14)
HumanCellAtlas/dcp2: docs/dcp2_operating_procedures.rst

.. |ne| replace:: :raw-html:`</blockquote>`

==========================
DCP/2 Operating Procedures
==========================

|nn| Text in a block quote (like this) is not normative. It was included
for informational purposes only and does not bear on the implementation.
|ne|

.. contents::

Changing the HCA metadata schema
================================

Changes to the HCA metadata schema, even seemingly minor ones, can have
significant impact on the individual components of the DCP system. The DCP
consortium therefore uses a review process that ensures all teams in the
consortium approve of a schema change **before** it 

- affects the master branch of the `metadata-schema`_ repository,

- is published on `schema.humancellatlas.org`_, and

- is used by metadata staged for import into the production instance of TDR

.. _schema.humancellatlas.org: https://schema.humancellatlas.org/a

.. _metadata-schema: https://github.com/HumanCellAtlas/metadata-schema

.. _schema-test-data: https://github.com/HumanCellAtlas/schema-test-data

.. _dcp2: https://github.com/HumanCellAtlas/dcp2

.. _DCP/2 system design: dcp2_system_design.rst

Sizing schema changes
---------------------

> A *schema change* is a set of related modifications to one or more of the 

Reviewer: I understand the sentiment of the paragraph but as part of the metadata team who generally makes the changes, I am not sure I would understand what changes can and can't be put into a single PR. I think it is fairly subjective. I also think given how lengthy and involved the update and review process is, that components could be slightly flexible with allowing changes to be grouped where they may not be explicitly related, but are unlikely to have much impact on downstream components. e.g. A minor schema change that adds an optional field that components can choose to ignore until they are able to support indexing it in the future.

Author: It is subjective but I think I give two good rules of thumb as a guideline. The current CGM PR started off with various things mixed in that ideally shouldn't have been. This is a ultimately a risk that authors have to take and judge. If a reviewer gets a kitchen sink PR, they'll reject it. Same if they can't see the tree for the forest because the PR is too small. If a PR groups unrelated changes A and B, it's possible that A is rejected so neither A and B gets accepted. If each change were filed separately, at least B would have been accepted. Feel free to suggest concrete changes that reflect the inherent subjectivity.

>I also think given how lengthy and involved the update and review process is

It is no more involved than the frustrating stalemate we've been operating under for the past four years where EBI creates facts and UCSC has to either scramble to catch up or throw in the gauntlet. Lets give this a chance.

JSONSchema documents in the `metadata-schema`_ repository. That set must be
chosen carefully. If a schema change *can* be made smaller by excluding a
particular change to a schema document without sacrificing the desired end-user
value, it *should* be made smaller, so as to not overwhelm the reviewers with a
mix of orthogonal concerns. On the other hand, the set of changes shoudn't be
too small either: all schema document changes that are necessary to achieve the
desired end-user value should be combined in one overall schema change. If
certain end user value is achieved in a series of many small schema changes,
reviewers will likely not be able to see the big picture until they review the
changes at the end of that series at which point it may be too late to reverse
course.

Some schema changes warrant an update to the `DCP/2 System Design`_
specification. Whether they do or not is established during the review process
of *pull requests* (PRs) against the `metadata-schema` and `schema-test-data`
repositories. Specification updates are done as a PR against the `dcp2`_
repository.

Schema change authorship
------------------------

> Anyone can author a PR. The PRs for one schema change don't all need to be by

Reviewer: I am not sure the best place to refer to it but anyone authoring a PR needs to follow the procedures in the documentation within the metadata-schema repository. I would also prefer that any changes to schema are done in consultation with the metadata-schema team as we have a large backlog of changes that we would like to prioritise.

Author: How would you change the wording?

the same author, as long as there are no competing and open PRs in the same
repository and for the same schema change. One person *should* author the PRs
against the `metadata-schema`_ and `dcp2`_ repositories. That person is referred
to as the *schema change author*. The schema change author can delegate
authorship of PRs to other members of the DCP consortium.

> Schema change procedure

Reviewer: The procedure misses the internal metadata team review of a PR that we do prior to opening for DCP review. I think we will need to review how we do this given the suggested procedure below.

Author: You could do the internal review on the develop branch PR.

-----------------------

The overall procedure for authoring and reviewing a schema change consists of
the following steps:

1)  If it is considered unnecessary to update the specification and a
    specification update wasn't explicitly requested, proceed to step 4.

2)  Prepare a PR against the `dcp2`_ repository, updating the `DCP/2 System
    Design`_.

3)  Initiate a `Pull request review`_ of your PR. Once the PR has been approved
    by the DCP, move on to the next step. Don't merge the PR just yet.

> 4)  Prepare a PR against the ``develop`` branch of the `metadata-schema`_

Reviewer: I will let @ESapenaVentura comment here but I think we are getting rid of the develop branch and merging directly to staging

Author: That sounds odd. Surely there must be a branch that corresponds to your development environment. This would rob us of the possibility to review multiple PRs in parallel.

Reviewer 2: @mshadbolt we are still merging to develop before merging to staging

Basically, the overall flow now is `many related feature branches` --> `dev` --> `staging` 

And the review by the downstream components happens in the PR from staging to master

>That sounds odd. Surely there must be a branch that corresponds to your development environment. This would rob us of the possibility to review multiple PRs in parallel.

Even with the above in mind, we still don't have the possibility to have parallel PRs for unrelated updates. Changes need to be merged into an environment to be published and therefore can't be in the feature branches only. We can only generate valid test data if it's published.

Author: > Even with the above in mind, we still don't have the possibility to have parallel PRs for unrelated updates.

You *can* have multiple PRs open against `develop`. That's what this SOP assumes.

Reviewer 2: Yep disregard the comment, I misread this for the step 6.

    repository. It is OK to perform this step concurrently with the preceding
    steps.

5)  Initiate a `Pull request review`_ of your PR. Once the PR has been approved
    by the DCP, move on to the next step. Don't merge the PR just yet.

    Note that the above steps can be performed for several independent schema
    changes in parallel.
   
6)  Merge the `metadata-schema`_ repository PR into the ``develop`` branch and
    from there into the ``staging`` branch.

7)  Prepare a PR against the `Test metadata`_ in the `schema-test-data`_
    repository. Don't merge the PR just yet.

8)  Initiate a `Pull request review`_ of your PR. Once the PR has been approved

>     by the DCP, move on to the next step. As part of the review, the Azul team

Reviewer: Is the Azul team the only consumer of the test data? i.e. the Azul unit tests cover all components? Asking because I am not sure if the full PR review process is needed for the test data or it is more for the schema change PR.

Author: I don't know of anyone else. The Azul unit tests do not cover all components  but there are three: Ingest, TDR and Azul/DB. TDR is mostly agnostic to metadata.

    updates their unit tests to validate the updated test metadata. If a test
    fails, they either improve the test or the team member reviewing the PR

>     requests the necessarty changes to the test metadata.

Reviewer: ```suggestion
    requests the necessary changes to the test metadata.
```

Author: Fixed.

> 9)  Prepare a PR against the schema repository, targetting the main branch.

Reviewer: ```suggestion
9)  Prepare a PR against the schema repository, targeting the master branch.
```

Author: Fixed.

> 10) Initiate a `Pull request review`_ of your PR. The review should be quick

Reviewer: So based on this process a single metadata schema PR may take up to 6 weeks to merge to master?
2 week review period for schemas to merge into staging (or more if changes are requested and review cycle is restarted)
2 week review of test data
2 week review of merge into master
Or if not the time periods above can we be explicit about the review periods for each review?

Author: The fastest possible is 1 week against develop, 1 week test data and staging. That's two weeks. This is without spec PR, and assuming no rejections and no dawdlers.

There is no upper limit. Each cycle could introduce new objections.

The section about the review process *is* explicit.

    because reviewers will already have seen the changes while reviewing the PR
    against the ``develop`` branch. Reviewers should not request any substantial
    changes at this stage unless last minute considerations absolutely require
    it. Once the PR has been approved by the DCP, move on to the next step.

11) Merge all PRs. You may squash and/or rebase a PR before merging it but that
    should not affect the diff between the base and head commits of the PR
    branch, except for housekeeping like resolving merge conflicts or adjusting
    the host name of schema URLs in PRs against the `schema-test-data`_
    repository. No semantic changes may be introduced to a PR after it has been
    approved by the DCP. All such housekeeping should be done on the PR branch
    prior to merging it.

Test metadata
-------------

Every change to the schemas in the `metadata-schema`_ repository must be
accompanied by a matching change to the test metadata in the `schema-test-data`_
repository. Both changes must be done as PRs following the `Pull request
review`_  process. 

> The test metadata should excercise all graph shapes in actual use. 

Reviewer: ```suggestion
The test metadata should exercise all graph shapes in actual use. 
```

Author: Fixed.

> The test metadata should excercise all graph shapes in actual use. 

Reviewer: Is it also expected that if we want to try out a new graph shape that we would make test data for it and have it approved before attempting export?

Author: Not as part of this document. Note the "should". But it would be the collegial and prudent thing to do. What counts is when does the end user get to see the new shape. And that is not until Azul is able to index it.

What constitutes a new shape is also subjective. Common sense applies.

The schema references (``describedBy``) in all metadata documents in the main
branch of `schema-test-data`_ repository must match the schemas on the

> `master` branch of the `metadata-schema`_ repository. Pull requests against the 

Reviewer: this sentence isn't complete

Author: I don't remember what I was going to write there. I'll remove it.

Author: Fixed

Pull request review
-------------------

> As a PR author

Reviewer: I assume you meant the terminology below since you spent some time defining it earlier. Or if not here because this process doesn't apply just to schema change I am not sure which responsibilities fall to the `schema change author`

```suggestion
As a *schema change author* 
```

Author: I do mean PR author. Note the

> The schema change author can delegate authorship of PRs to other members of the DCP consortium.

in that section.

Reviewer 3: I think this is just framing Marion's question differently -- 1. whats a scenario of when someone is the schema change author and not the PR author -- and what is the value of separating those roles? 2. if one is a schema change author and delegated authorship of PRs to other members of the DCP -- then what is the role/responsibility of the schema change author in that scenario?

Author: I was trying to accommodate what I think I observed during the organic CGM effort. Different people were authoring different PRs. I'd be curious to what @ESapenaVentura and @amnonkhen think about this.

Reviewer 2: If I understood it correctly, we're talking here about responsibilities, right? e.g. A PR may be opened by person A but person B may be the schema change author and therefore is responsible for requests.

I do think that this distinction is useful, especially in cases where e.g. a person starts a PR for a schema change and then due to some unexpected circumstances is not reliable to lead the effort of responding to requests/changes for the next 2 weeks.

Most of the times person A will be both schema change author and PR author, but preparing for different case scenarios seems fine to me.

It may need some clarification if the distinction between both roles is not clear to the readers, though.

Author: As written, the spec doesn't accommodate the possibility of PR authors changing mid-PR. It only accommodates the possibility that the PRs involved in a schema change are authored by different people. It requires that one person is responsible for the schema change as whole. This is also often referred to as a *sponsor*. We need one person to pull the strings and be accountable for every schema change.

Author: Fixed

1)  Announce the PR on the `#dcp2` channel on Slack, @-mentioning all
    `Designated reviewers`_.

> 2)  Request a review from all `Designated reviewers`_.

Reviewer: ```suggestion
2)  Reviews are automatically requested by notifying the `Designated reviewers`_ listed in the pullapprove configuration file.
```

Author: We can't refer to two lists because they could be inconsistent. PA is only used in the schema repo and I think this document should be self-contained, especially when it comes to the very important list of reviewers.

Reviewer 2: With self-containing in mind, agree that there is no use in listing how does the PR author identify the designated reviewers

Author: won't fix

3)  Wait one week.

4)  If this is the first review cycle, remind about the PR on the ``#dcp2``
    channel on Slack, @-mentioning requested reviewers that haven't yet
    reviewed the PR.

5)  If two weeks have passed since step 1 and there are no binding reviews
    requesting changes, the PR has been approved by the DCP.

6)  Otherwise, if this is a PR against the `metadata-schema`_ or
    `schema-test-data`_ repositories and a reviewer requests that you first
    update the DCP/2 system design specifcation first, close this PR and
    proceed with step 2 in section `Changing the HCA metadata schema`_.

7)  Otherwise, respond to every review comment either by making a source code
    change that you think appropriately addresses the comment or by replying
    with a comment explaining your opposition to the change or asking for more
    information. When addressing a set of related review comments with a source
    code change, try to commit that change separately from those addressing
    other related sets of review comments. Avoid constantly squashing the PR
    branch unless doing so helps reviewers to better understand the branch
    history.

> 8)  Start another cycle by requesting a review from the reviewers currently on

Reviewer: so with each change that is made the review process gets extended by another 2 weeks?

Author: No, just one week. The two weeks are from the overall beginning (step 1), they don't restart every cycle. I want to give everyone at least two weeks to chime in on the PR overall, not every cycle.

I'll add language to allow for the 1 week "everyone approves" scenario.

Author: Fixed

delegate but that should be kept to a minimum and ideally be done outside of
Github e.g., Slack or E-Mail.

A delegate reviewer can only delegate back to the designated reviewer that named
them, and only after the PR author rerequested a review from the delegate.

Only comments and reviews by mediators, designated reviewers or their delegate
are binding. Other reviews should be `dismissed`_ by the author. Other comments
can be ignored.
  
Reviews by designated reviewers or their delegate can be `dismissed`_ by the
`Mediators`_.

At no time during the life-time of the PR can there be more reviewers listed on
the PR than there are entries in the lists of `Designated reviewers`_ and
`Mediators`_.

Discourse
~~~~~~~~~

Review comments and replies by authors should be kept brief. Typically, review
comments made during a review cycle are addressed by the author and marked
resolved by the reviewer during the next cycle. Any disagreement that cannot be
resolved in two cycles should be discussed in a conference call to which the the
PR author invites the `Mediators`_ and all reviewers currently on the PR.

Designated reviewers
~~~~~~~~~~~~~~~~~~~~

- Claire (EBI)

> - Marion (EBI)

Reviewer: I don't think it is worth listing me here, I will not be part of the DCP in a few weeks

```suggestion
```

Author: I feel that two EBI people should be listed. Who should I list instead?

Author: Elsewhere you nominated @amnonkhen so I went with him.

Author: Fixed

Once a designated reviewer delegated the review, none of the designated
reviewer's comments or reviews are binding. It's acceptable to direct the
delegate but that should be kept to a minimum and ideally be done outside of
Github e.g., Slack or E-Mail.

A delegate reviewer can only delegate back to the designated reviewer that named
them, and only after the PR author rerequested a review from the delegate.

Only comments and reviews by mediators, designated reviewers or their delegate
are binding. Other reviews should be `dismissed`_ by the author. Other comments
can be ignored.
  
Reviews by designated reviewers or their delegate can be `dismissed`_ by the
`Mediators`_.

At no time during the life-time of the PR can there be more reviewers listed on
the PR than there are entries in the lists of `Designated reviewers`_ and
`Mediators`_.

Discourse
~~~~~~~~~

Review comments and replies by authors should be kept brief. Typically, review
comments made during a review cycle are addressed by the author and marked
resolved by the reviewer during the next cycle. Any disagreement that cannot be
resolved in two cycles should be discussed in a conference call to which the the

> PR author invites the `Mediators`_ and all reviewers currently on the PR.

Author: Meetings should occur earlier and be standing.

Author: Mention two types of approval "don't care" and "do care but now its ok".

Author: Won't fix now due to time constraints.

changes collectively to ensure that all teams in the consortium approve of a
schema change **before** it 

- affects the master branch of the `metadata-schema`_ repository,

- is published on `schema.humancellatlas.org`_, or

- is used by metadata staged for import into the production instance of TDR

.. _schema.humancellatlas.org: https://schema.humancellatlas.org/a

.. _metadata-schema: https://github.com/HumanCellAtlas/metadata-schema

.. _schema-test-data: https://github.com/HumanCellAtlas/schema-test-data

.. _dcp2: https://github.com/HumanCellAtlas/dcp2

.. _DCP/2 system design: dcp2_system_design.rst

Impact of Schema changes
-------------------------

Changes to the HCA metadata schema, including seemingly minor ones, can have
significant impact on the individual components of the DCP system. Even the
addition of a property can, in some case, require significant work in
metadata consumers like the Data Browser, especially if those consumers are
required to interpret the added property. It would be difficult to explicitly
define the criteria for classifing schema changes by impact. Instead, the
impact is determined individually for every schema change. Changes that are

> deemed of low impactare fast-tracked through a lighter-weight review process.

Reviewer 4: ```suggestion
deemed of low impact are fast-tracked through a lighter-weight review process.
```

Author: fixed

Sizing schema changes
---------------------

> A *schema change* is a carefully seleced set of related modifications to one

Reviewer 5: ```suggestion
A *schema change* is a carefully selected set of related modifications to one
```

Author: fixed

on during the review process.

Responsiveness to submitter requirements
----------------------------------------

Schema changes are often motivated by the need to capture novel information
about the experimental design, phenotypes or sample collection process as
submitted by a participating laboratory, in a form that the current schema is
not able to sufficiently capture. Participating labs often don't have the
resources or tolerance for a prolonged submission process and are likely to
walk away if it takes too long to incorporate their submission into the
DCP/2.

The DCP/2 therefore needs to balance the celerity with which it adopts schema
changes on one hand against the effort needed to make all components
compatible with those changes on the other. The earlier a proposed change
faces scrutiny by the entire DCP/2 consortium, the better will the consortium
be able to negotiate the right balance, simply because the change will not
yet have permeated through the system, making it easier to make adujustments
to the proposed schema changes.

Review process overview
-----------------------

The process of reviewing schema changes utilizes Github's pull request

> (PR) feature. Every schema change begins with a PR against the `develop`

Reviewer 2: ```suggestion
(PR) feature. Every schema change begins with a PR against the `staging`
```

We are getting rid of the develop and integration environments in the schema repository, as they do not add any value and delay the release process

Author: Will you remove the `develop` and `integration` branches? If no, will there ever be commits to those branches in the future?

Reviewer 2: Yes! we are planning to remove them. Currently the schema publisher only pushes to the schema prod and staging environment, so the remaining should be removed soon. It's a matter of having someone go through the task of deleting the branches and updating the docs!

Author: Fixed

> branch of the `metadata-schema`_ repository. This is were the DCP/2

Reviewer 5: ```suggestion
branch of the `metadata-schema`_ repository. This is where the DCP/2
```

Author: Fixed

consortium determines

* the impact of the change

* whether the schema change is `sized <Sizing schema changes>`_ correctly

* whether the change requires updating the specification in the `dcp2`_ repository

* whether it requires updating the test metadata in the `schema-test-data`_
  repository for other components to code unit tests against

If a change requires specification changes, the `metadata-schema`_ PR review is
suspended and a PR is filed against the `dcp2`_ repository and reviewed there.
Once that PR is approved, the `metadata-schema`_ PR is revised and reviewed
again.

Schema changes that are sized correctly and have negligible impact to other
components, are considered adopted by the DCP/2 as soon as the PR against the

> `develop` branch is approved. No further reviews of PRs against other branches or

Reviewer 2: ```suggestion
`staging` branch is approved. No further reviews of PRs against other branches or
```
See comment above about deprecation of develop and integration environments

Author: Fixed

repositories are required.

Schema changes that are sized correctly, and that are of some impact to other
components, may require more involved testing. There are two ways of testing
schema changes: 

> A)  (Meta)data exhibiting the proposed changes are first staged and then

Reviewer 2: ```suggestion
A) (Meta)data exhibiting the proposed schema changes are first staged and then
```

Author: fixed

    imported into the `dev` instance of TDR after which downstream components
    use the resulting `dev` snapshot to test the code changes that are required
    to support the schema change. The Ingest component can only populate

>     staging areas from its staging instance, so the `develop` PR must be

Reviewer 2: ```suggestion
staging areas from its staging instance, so the `staging` PR must be
```

Author: Fixed

>     approved(under the provision that the schema change may need to be amended

Reviewer 5: ```suggestion
    approved (under the provision that the schema change may need to be amended
```

Author: fixed

    by another PR) and the changes are promoted to the staging branch.

B) (Meta)data exhibiting the proposed schema changes are committed to a feature
    branch of the `schema-test-data`_ repository, and a PR is filed for that
    branch. The `schema-test-data`_ repository can only populated using the

>     staging instance of the Ingest component, so the PR against the `develop`

Reviewer 5: ```suggestion
    staging instance of the Ingest component, so the PR against the `staging`
```

Author: fixed

    branch of the `metadata-schema`_ repository must be approved (under the
    condition that the schema change may need to be amended in another PR)
    and the changes are promoted to the staging branch.

The reviewers of a PR can request either testing strategy. Strategy A will
likely be requested for for most changes with some impact on other components.

In some cases a lab submission not only involves a schema change but also
introduces a novel way of linking the metadata enties into subgraphs. In those

> cases, reviewers are likely to request testing strategies B. A request for

Reviewer 5: ```suggestion
cases, reviewers are likely to request testing strategy B. A request for
```

Author: fixed

> either strategy should be reasonably motivated. Possible reasons include 

Reviewer 5: ```suggestion
either strategy should be reasonably motivated. Possible reasons include: 
```

Author: I don't think so. I checked CMOS. If you care about this, post supporting evidence.

- the need to not only review schema changes but also review how those changes
  affect actual metadata documents

- the need to test future, unrelated code changes against the test metadata, so
  as to make sure that those future changes don't introduce regressions.
  Especially graph changes fall into that category.

Schema changes that are sized incorrectly need to be revised.

Schema change authorship
------------------------

Anyone can author a schema change. The PRs for one schema change don't all need
to be by the same author, as long as there are no competing PRs pending in the
same repository, for the same schema change. One person *should* author the PRs
against the `metadata-schema`_ and `dcp2`_ repositories. That person is
referred to as the *schema change author*. The schema change author can
delegate authorship of PRs to other members of the DCP consortium.

Pull request reviews
--------------------

Pull requests against the `dcp2`_, the `metadata-schema`_ or the

> `schema-test-data` repositories are reviewed and approved in exactly the same

Reviewer 2: ```suggestion
`schema-test-data`_ repositories are reviewed and approved in exactly the same
```

Author: fixed

way, provided they are involved in a semantic change to either the DCP/2
specification, a DCP/2 standard operating procedure or a metadata schema
change:

As a PR author

1)  Announce the PR on the ``#dcp2`` channel on Slack, @-mentioning all
    `Designated reviewers`_.

2)  Request a review from all `Designated reviewers`_.

3)  Wait one week.

4)  If this is the first review cycle, remind about the PR on the ``#dcp2``
    channel on Slack, @-mentioning requested reviewers that haven't yet
    reviewed the PR.

5)  If either

    a)  all reviewers approve of the PR without conditions or

    b)  two weeks have passed since step 1 and there are no binding reviews
        requesting changes,

    merge the PR. You are done unless any of the approvals are conditional. A

>     conditional approval is one that's dependent on successful testing using

Reviewer 2: How do we identify conditional approvals? Are we expecting a comment on the PR, a successful review with a comment?

Author: Via comment. I'll clarify this.

Author: Fixed

>     one of the strategies mentioned in `Review process overview_`. If testing

Reviewer 5: ```suggestion
    one of the strategies mentioned in `Review process overview`_. If testing
```

Author: Fixed.

    the requested amendmends and start at step 1.

6)  If a reviewer requests that you first update the DCP/2 system design
    specifcation or standard operating procedures first, suspend work on this
    PR and open another PR against the `dcp2`_ repository. Start at step 1
    there. Once that PR has been approved, update this PR branch with any
    follow-up changes resulting from the `dcp2`_ PR review process and resume
    this PR at step 1.

7)  Otherwise, respond to every review comment either by making a source code
    change that you think appropriately addresses the comment or by replying
    with a comment explaining your opposition to the change or asking for more
    information. When addressing a set of related review comments with a source
    code change, try to commit that change separately from those addressing
    other related sets of review comments. Avoid constantly squashing the PR
    branch unless doing so helps reviewers to better understand the branch
    history.

8)  Start another cycle by requesting a review from the reviewers currently on
    the PR, even those that already approved the PR or just commented on it.
    Proceed to step 3.

> Steps 1, 2 and 8 must be done on a weekday.

Reviewer 3: good catch :)

delegate but that should be kept to a minimum and ideally be done outside of
Github e.g., Slack or E-Mail.

A delegate reviewer can only delegate back to the designated reviewer that named
them, and only after the PR author rerequested a review from the delegate.

Only comments and reviews by mediators, designated reviewers or their delegate
are binding. Other reviews should be `dismissed`_ by the author. Other comments
can be ignored.
  
Reviews by designated reviewers or their delegate can be `dismissed`_ by the
`Mediators`_.

At no time during the life-time of the PR can there be more reviewers listed on
the PR than there are entries in the lists of `Designated reviewers`_ and
`Mediators`_.

Discourse
~~~~~~~~~

Review comments and replies by authors should be kept brief. Typically, review
comments made during a review cycle are addressed by the author and marked
resolved by the reviewer during the next cycle. Any disagreement that cannot be
resolved in two cycles should be discussed in a conference call to which the the
PR author invites the `Mediators`_ and all reviewers currently on the PR.

Designated reviewers
~~~~~~~~~~~~~~~~~~~~

> - Claire aka ``@clairerye`` (EBI)

Reviewer 2: ```suggestion
- Enrique aka ``@ESapenaVentura`` (EBI)
```

Author: Fixed

    merge the PR. You are done unless any of the approvals are conditional. A
    conditional approval is one that's dependent on successful testing using
    one of the strategies mentioned in `Review process overview`_. The
    condition has to be specified in the summary comment on approval. If

>     testing fails, a reviewer may request amendmends to your changes. These

Reviewer 5: ```suggestion
    testing fails, a reviewer may request amendments to your changes. These
```

    requests are made as comments to the original, now merged PR. Open

>     another PR with the requested amendmends and start at step 1.

Reviewer 5: ```suggestion
    another PR with the requested amendments and start at step 1.
```
